main/eval/src/mill/eval/KeyLock.scala

```diff
@@ -0,0 +1,26 @@
+package mill.eval
+
+private class KeyLock[K]() {
+  private[this] val lockKeys = new java.util.HashSet[K]()
+
+  def lock(key: K, onCollision: Option[() => Unit] = None): AutoCloseable = {
+    lockKeys.synchronized {
+      var shown = false
+      while (!lockKeys.add(key)) {
```

Comment:
Does this need to be a `while`? I thought we're inside `lockKeys.synchronized`, there should be no need to "try again" since nobody else else going to be able to mutate `lockKeys`

Reply:
I think yes. At least, the Javadoc of the `Object.wait` method suggests to do so:

> As in the one argument version, interrupts and spurious wakeups are possible, and this method should always be used in a loop:
>
> ```java
> synchronized (obj) {
>     while (<condition does not hold>)
>         obj.wait();
>     ... // Perform action appropriate to condition
> }
> ```